private/buf/bufwork/bufwork.go

```diff
@@ -141,13 +143,51 @@ func BuildOptionsForWorkspaceDirectory(
 		// Managed Mode, which supports module-specific overrides.
 		bufmodulebuild.WithModuleIdentity(moduleConfig.ModuleIdentity),
 	}
-	if len(externalDirOrFilePaths) == 0 {
+	if len(externalDirOrFilePaths) == 0 && len(excludeDirOrFilePaths) == 0 {
 		return buildOptions, nil
 	}
 	// We know that if the file is actually buf.work for legacy reasons, this will be wrong,
 	// but we accept that as this shouldn't happen often anymore and this is just
 	// used for error messages.
 	workspaceID := filepath.Join(normalpath.Unnormalize(relativeRootPath), ExternalConfigV1FilePath)
+	// Note that subDirRelPaths can be empty. If so, this represents
+	// the case where externalDirOrFilePaths were provided, but none
+	// matched.
+	subDirRelPaths, err := externalPathsToSubDirRelPaths(
+		workspaceID,
+		relativeRootPath,
+		subDirPath,
+		externalDirOrFilePaths,
+	)
```

Comment:
Big fan of this refactor - this is much easier to read with `externalPathsToSubDirRelPaths` now 👏